src/main/scala/is/hail/variant/Genotype.scala

```diff
@@ -787,6 +807,156 @@ class GenericGenotype(private val _gt: Int,
       Some(Genotype.phredToDosage(_px))
 }
 
+class MutableGenotype(nAlleles: Int) extends Genotype {
+
+  private var _gt: Int = 0
+  private val _ad: Array[Int] = Array.ofDim[Int](nAlleles)
+  private var _dp: Int = 0
+  private var _gq: Int = 0
+  private val _px: Array[Int] = Array.ofDim[Int](triangle(nAlleles))
+  var _fakeRef: Boolean = false
+  var _isDosage: Boolean = false
+
+  if (isDosage) {
+    require(_gq == -1)
+    if (_px(0) == -1)
+      require(_gt == -1)
+    else {
+      require(_px.sum == 32768)
+      require(_gt == Genotype.gtFromLinear(_px).getOrElse(-1))
+    }
+  }
+
+  def unboxedGT: Int = _gt
+  def fakeRef: Boolean = _fakeRef
+  def isDosage: Boolean = _isDosage
+
+  def ad: Option[Array[Int]] = if (_ad(0) >= 0) Some(_ad) else None
```

Comment:
I wouldn't reuse `_ad(0)` this way.  Just make a `private var hasAD: Boolean = false`.